File: 01/main.py
#!/usr/bin/python

import argparse


def find_pair_internal(numbers, target):
    for i in range(len(numbers) - 1):
        current = numbers[i]
        for j in range(i + 1, len(numbers)):
            check = numbers[j]
            if current + check == target:
                return (current, check)
    return (None, None)


def find_pair(numbers):
    a, b = find_pair_internal(numbers, 2020)
    if a is None:
        print('Could not find result :(')
        return
    print('%s + %s = 2020' % (a, b))
    print('Result: %s' % (a * b))


def find_triplet(numbers):
    for i in range(len(numbers) - 2):
        current = numbers[i]
        a, b = find_pair_internal(numbers[i + 1:], 2020 - current)
        if a is None:
            continue
        print('%s + %s + %s = 2020' % (current, a, b))
        print('Result: %s' % (current * a * b))
        return
    print('Could not find result :(')


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('file')
    args = parser.parse_args()

    with open(args.file, 'r') as input_file:
        numbers = [int(l.strip()) for l in input_file.readlines()]

    # find_pair(numbers)
    find_triplet(numbers)


if __name__ == "__main__":
    main()

File: 01/src/main.rs
use anyhow::{bail, Result};
use std::fs::File;
use std::io::{prelude::*, BufReader};
use structopt::StructOpt;

fn find_pair_internal(numbers: &[u32], target: u32) -> Result<(u32, u32)> {
    for (i, number) in numbers.iter().enumerate() {
        for candidate in numbers[(i + 1)..numbers.len()].iter() {
            if number + candidate == target {
                return Ok((*number, *candidate));
            }
        }
    }
    bail!("No pair found matching {}", target);
}

fn find_pair(numbers: &[u32]) -> Result<u32> {
    let (a, b) = find_pair_internal(numbers, 2020)?;
    Ok(a * b)
}

fn find_triplet(numbers: &[u32]) -> Result<u32> {
    for (i, number) in numbers[0..(numbers.len() - 2)].iter().enumerate() {
        if let Ok((a, b)) = find_pair_internal(&numbers[(i + 1)..numbers.len()], 2020 - number) {
            return Ok(a * b * number);
        }
    }
    bail!("Could not find triplet totalling 2020");
}

#[derive(Debug, StructOpt)]
pub struct Flags {
    #[structopt(short = "i", long = "file")]
    pub file: String,
    #[structopt(short = "n", long = "next")]
    pub next: bool,
}

fn main() -> Result<()> {
    let flags = Flags::from_args();
    let file = File::open(flags.file)?;
    let numbers: Vec<u32> = BufReader::new(file)
        .lines()
        .map(|l| {
            l.expect("Could not read line")
                .parse()
                .expect("Could not parse number")
        })
        .collect();
    let result = if flags.next {
        find_triplet(&numbers)?
    } else {
        find_pair(&numbers)?
    };
    println!("{:?}", result);
    Ok(())
}

#[cfg(test)]
mod tests {
    use crate::{find_pair, find_triplet};

    #[test]
    fn test_find_pair() {
        assert_eq!(
            find_pair(&vec![1721, 979, 366, 299, 675, 1456]).unwrap(),
            514579
        );
    }

    #[test]
    fn test_find_triplet() {
        assert_eq!(
            find_triplet(&vec![1721, 979, 366, 299, 675, 1456]).unwrap(),
            241861950
        );
    }
}
